### ISSUE TITLE
adapters do not have an opportunity to do long-running activities such as add to room

### ISSUE THREAD

**Issue by bytenik** [socketio/socket.io#3662]
### You want to:

* [] report a *bug*
* [x] request a *feature*

### Current behaviour

socket.io calls functions on the adapter, and then immediately invokes its own internal callback function with an argument of `null`.

### Expected behaviour

socket.io should pass a callback function to adapter functions and then use that to trigger callback to the rest of the socket.io stack.

### Other information (e.g. stacktraces, related issues, suggestions how to fix)

This is particularly critical for the adapter I am developing that works with AWS SQS/SNS -- https://github.com/thinkalpha/socket.io-sqs
Specifically, actions like `addAll` can't be properly handled in time; e.g. if you `addAll` a socket to room xxx and then immediately `.to(xxx).emit(...)`, your message doesn't necessarily end up back at that socket.

**Comment by darrachequesne** [comment]
The `adapter` actually had a callback in the previous versions (included in Socket.IO v2), see [here](https://github.com/socketio/socket.io-adapter/blob/1.1.2/index.js).

We've removed it since it was not needed anymore by the Redis adapter, but it seems other adapters may benefit from this. We could add it back in the form of a Promise, would it suit your use case?

**Comment by bytenik** [comment]
Yes, making it promise-based would be ideal!

**Comment by bytenik** [comment]
Incorporating #3661, #3660, and #3659 into this issue.

**Comment by bytenik** [comment]
I've worked around some of these issues in a really hacky way by making a factory for the adapter class and then firing callbacks on arguments passed into the factory, but its definitely not ideal.

**Comment by darrachequesne** [comment]
@bytenik I've updated the adapter class (https://github.com/socketio/socket.io-adapter/commit/2e023bf2b651e543a34147fab19497fbdb8bdb72) and the socket#join() method (https://github.com/socketio/socket.io/commit/129c6417bd818bc8b4e1b831644323876e627c13).

I'm not sure about the public API for the Adapter#init() and Adapter#close() methods though. Should those methods be called by the end users directly?

```js
const server = require("http").createServer();
const io = require("socket.io")(server);

await this.of("/").adapter.init();

server.listen(3000);

// and then...
server.close();
await this.of("/").adapter.close();
```

**Comment by bytenik** [comment]
@darrachequesne Awesome, that looks great.

Right now the Adapter constructor is passed in, not a constructed adapter. socket.io constructs the adapter. So, there's no good way for the end user to call init or close. The lifespan is controlled by the io instance.

**Comment by darrachequesne** [comment]
Yes you are right. Currently, there is one Adapter per namespace, and they are constructed when:

- a namespace is created, with `io.of("/my-namespace")`
- the adapter constructor is updated (for all existing namespaces) with `io.adapter(CustomAdapter)`

But we could change that behavior, though I'm not really sure about the implications.

**Comment by bytenik** [comment]
@darrachequesne I don't think that there's anything inherently wrong with socket.io controlling the adapter lifecycle. I just think that in that case, the adapter needs an opportunity to start long running processes (init) and shutdown long running processes (close). So, both could return promises and socket.io would call those functions and then block until they resolve.

**Comment by bytenik** [comment]
Hey, I just got back to trying to implement these changes on my end. Is `init` ever being called? I can't find a call to it, and my debug messages indicate it isn't being hit.

**Comment by darrachequesne** [comment]
You are right, the `Adapter#init` method is not currently called.

We could call it here:

https://github.com/socketio/socket.io/blob/1faa7e3aea1414ec814aa935021356e8ed2b054c/lib/namespace.ts#L79-L81

But we cannot await a constructor (the `initAdapter` method is called under the hood):

```js
const io = new Server({
  adapter: myCustomAdapter
});
```

Is there any problem with manually calling `init()`?

```js
const io = new Server({
  adapter: myCustomAdapter
});
await io.of("/").adapter.init();
```

I'm open to suggestions on this.

**Comment by darrachequesne** [comment]
For future readers:

- `init()` and `close()` were added to the Adapter class
- `adapter.init()` is called when creating the namespace (but cannot be awaited)
- `adapter.close()` is called when closing the server

Commits:

- https://github.com/socketio/socket.io-adapter/commit/2e023bf2b651e543a34147fab19497fbdb8bdb72 (`socket.io@3.0.0`, Oct 2020)
- https://github.com/socketio/socket.io/commit/bf64870957e626a73e0544716a1a41a4ba5093bb  (`socket.io@4.7.5`, Mar 2024)
- https://github.com/socketio/socket.io/commit/f3e1f5ebdf59158d0c8d1e20f8230275617fb355 (`socket.io@4.8.2`, Dec 2025)
